File: backend/app/job_vector.py
"""Build job_vector scores from a job description and skill keywords."""

from __future__ import annotations

import json
import re
from typing import Any


def mention_score(count: int) -> float:
    if count <= 0:
        return 0.0
    if count == 1:
        return 1.0
    if count == 2:
        return 1.5
    if count == 3:
        return 2.0
    return 2.5


def count_mentions(job_description: str, keyword: str) -> int:
    text = (keyword or "").strip()
    if not text:
        return 0
    pattern = re.compile(re.escape(text), re.IGNORECASE)
    return len(pattern.findall(job_description or ""))


def build_job_vector(job_description: str, keywords: list[str]) -> list[float]:
    return [mention_score(count_mentions(job_description, keyword)) for keyword in keywords]


def build_job_vector_weighted(
    job_description: str,
    entries: list[tuple[str, float]],
) -> list[float]:
    scores: list[float] = []
    for keyword, weight in entries:
        score = mention_score(count_mentions(job_description, keyword))
        multiplier = float(weight if weight is not None else 1.0)
        scores.append(score * multiplier)
    return scores


def _entry_to_keyword(entry: Any) -> str:
    if entry is None:
        return ""
    if isinstance(entry, (str, int, float)):
        return str(entry).strip()
    if isinstance(entry, dict):
        item = entry.get("item")
        if item is not None and str(item).strip():
            return str(item).strip()
        keyword = entry.get("keyword")
        if keyword is not None and str(keyword).strip():
            return str(keyword).strip()
    return ""


def _normalize_keyword_json_text(text: str) -> str:
    """Quote bare keys so {item : "react", weight:1} becomes valid JSON."""
    return re.sub(
        r"([{,]\s*)([A-Za-z_][A-Za-z0-9_]*)\s*:",
        r'\1"\2":',
        text.strip(),
    )


def _extract_items_by_regex(text: str) -> list[str]:
    return [
        value.strip()
        for value in re.findall(r'\bitem\s*:\s*"([^"]+)"', text, flags=re.IGNORECASE)
        if value.strip()
    ]


def parse_keyword_weight_entries(raw: Any, *, default_weight: float = 1.0) -> list[tuple[str, float]]:
    """Normalize legacy keyword payloads into (keyword, weight) pairs."""
    if raw is None:
        return []

    if isinstance(raw, list):
        entries: list[tuple[str, float]] = []
        for item in raw:
            if isinstance(item, dict):
                keyword = _entry_to_keyword(item)
                if not keyword:
                    continue
                weight_raw = item.get("weight")
                weight = (
                    default_weight
                    if weight_raw is None
                    else float(weight_raw)
                )
                entries.append((keyword, weight))
            else:
                keyword = _entry_to_keyword(item)
                if keyword:
                    entries.append((keyword, default_weight))
        return entries

    if isinstance(raw, dict):
        nested = raw.get("items") or raw.get("keywords")
        if isinstance(nested, list):
            return parse_keyword_weight_entries(nested, default_weight=default_weight)
        keyword = _entry_to_keyword(raw)
        return [(keyword, default_weight)] if keyword else []

    text = str(raw).strip()
    if not text:
        return []

    try:
        return parse_keyword_weight_entries(json.loads(text), default_weight=default_weight)
    except json.JSONDecodeError:
        pass

    try:
        return parse_keyword_weight_entries(
            json.loads(_normalize_keyword_json_text(text)),
            default_weight=default_weight,
        )
    except json.JSONDecodeError:
        pass

    from_regex = _extract_items_by_regex(text)
    if from_regex:
        return [(value, default_weight) for value in from_regex]

    return [(text, default_weight)]


def parse_keyword_entries(raw: Any) -> list[str]:
    """
    Normalize skill.keyword into keyword strings.

    Expected shape: [{ "item": "react", "weight": 1 }, ...]
    Also accepts lenient unquoted-key text used in the UI.
    """
    if raw is None:
        return []
    if isinstance(raw, list):
        return [value for value in (_entry_to_keyword(item) for item in raw) if value]
    if isinstance(raw, dict):
        items = raw.get("items") or raw.get("keywords")
        if isinstance(items, list):
            return [value for value in (_entry_to_keyword(item) for item in items) if value]
        single = _entry_to_keyword(raw)
        return [single] if single else []

    text = str(raw).strip()
    if not text:
        return []

    try:
        return parse_keyword_entries(json.loads(text))
    except json.JSONDecodeError:
        pass

    try:
        return parse_keyword_entries(json.loads(_normalize_keyword_json_text(text)))
    except json.JSONDecodeError:
        pass

    from_regex = _extract_items_by_regex(text)
    if from_regex:
        return from_regex

    return [text]

File: frontend/src/utils/jobVector.js
/**
 * Build job_vector scores from a job description and skill keywords.
 *
 * Keyword source: skill.keyword JSON → array of { item, weight } → use item.
 * Match: case-insensitive substring include.
 * Scoring:
 *   0 mentions → 0
 *   1 → 1.0
 *   2 → 1.5
 *   3 → 2.0
 *   4+ → 2.5
 */

export function mentionScore(count) {
  if (count <= 0) return 0;
  if (count === 1) return 1.0;
  if (count === 2) return 1.5;
  if (count === 3) return 2.0;
  return 2.5;
}

function escapeRegExp(value) {
  return String(value).replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function countMentions(jobDescription, keyword) {
  const text = String(keyword || '').trim();
  if (!text) return 0;
  const pattern = new RegExp(escapeRegExp(text), 'gi');
  const matches = String(jobDescription || '').match(pattern);
  return matches ? matches.length : 0;
}

export function buildJobVector(jobDescription, keywords = []) {
  return keywords.map((keyword) => mentionScore(countMentions(jobDescription, keyword)));
}

function entryToKeyword(entry) {
  if (entry == null) return '';
  if (typeof entry === 'string' || typeof entry === 'number') {
    return String(entry).trim();
  }
  if (typeof entry === 'object') {
    if (entry.item != null && String(entry.item).trim()) {
      return String(entry.item).trim();
    }
    if (entry.keyword != null && String(entry.keyword).trim()) {
      return String(entry.keyword).trim();
    }
  }
  return '';
}

/** Turn JS-object-like skill keyword text into JSON-parseable text when needed. */
function normalizeKeywordJsonText(text) {
  let normalized = String(text || '').trim();
  if (!normalized) return '';
  // Quote bare keys: {item : "react", weight:1} → {"item":"react","weight":1}
  normalized = normalized.replace(
    /([{,]\s*)([A-Za-z_][A-Za-z0-9_]*)\s*:/g,
    '$1"$2":'
  );
  return normalized;
}

function extractItemsByRegex(text) {
  const keywords = [];
  const pattern = /\bitem\s*:\s*"([^"]+)"/gi;
  let match = pattern.exec(text);
  while (match) {
    const value = String(match[1] || '').trim();
    if (value) keywords.push(value);
    match = pattern.exec(text);
  }
  return keywords;
}

export function parseKeywordEntries(raw) {
  if (raw == null) return [];

  if (Array.isArray(raw)) {
    return raw.map(entryToKeyword).filter(Boolean);
  }

  if (typeof raw === 'object') {
    if (Array.isArray(raw.items)) {
      return raw.items.map(entryToKeyword).filter(Boolean);
    }
    if (Array.isArray(raw.keywords)) {
      return raw.keywords.map(entryToKeyword).filter(Boolean);
    }
    const single = entryToKeyword(raw);
    return single ? [single] : [];
  }

  const text = String(raw).trim();
  if (!text) return [];

  try {
    return parseKeywordEntries(JSON.parse(text));
  } catch {
    /* try lenient key quoting */
  }

  try {
    return parseKeywordEntries(JSON.parse(normalizeKeywordJsonText(text)));
  } catch {
    /* fall through to regex */
  }

  const fromRegex = extractItemsByRegex(text);
  if (fromRegex.length) {
    return fromRegex;
  }

  return [text];
}

export function flattenSkillKeywords(skills = []) {
  const keywords = [];
  skills.forEach((skill) => {
    keywords.push(...parseKeywordEntries(skill.keyword));
  });
  return keywords;
}
